In [21]:
# Useful starting lines
%matplotlib inline
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
from implementations import *
%load_ext autoreload
%autoreload 2

The autoreload extension is already loaded. To reload it, use:
  %reload_ext autoreload


## Load the training data into feature matrix, class labels, and event ids:

In [3]:
from proj1_helpers import *
DATA_TRAIN_PATH = "../data/train.csv" 
y, tX, ids = load_csv_data(DATA_TRAIN_PATH)
df= pd.read_csv(DATA_TRAIN_PATH)
df.head()

Id Prediction  DER_mass_MMC  DER_mass_transverse_met_lep  DER_mass_vis  \
0  100000          s       138.470                       51.655        97.827   
1  100001          b       160.937                       68.768       103.235   
2  100002          b      -999.000                      162.172       125.953   
3  100003          b       143.905                       81.417        80.943   
4  100004          b       175.864                       16.915       134.805   

   DER_pt_h  DER_deltaeta_jet_jet  DER_mass_jet_jet  DER_prodeta_jet_jet  \
0    27.980                  0.91           124.711                2.666   
1    48.146               -999.00          -999.000             -999.000   
2    35.635               -999.00          -999.000             -999.000   
3     0.414               -999.00          -999.000             -999.000   
4    16.405               -999.00          -999.000             -999.000   

   DER_deltar_tau_lep  ...  PRI_met_phi  PRI_met_sumet  PRI_jet_num  \
0               3.064  ...       -0.277        258.733            2   
1               3.473  ...       -1.916        164.546            1   
2               3.148  ...       -2.186        260.414            1   
3               3.310  ...        0.060         86.062            0   
4               3.891  ...       -0.871         53.131            0   

   PRI_jet_leading_pt  PRI_jet_leading_eta  PRI_jet_leading_phi  \
0              67.435                2.150                0.444   
1              46.226                0.725                1.158   
2              44.251                2.053               -2.028   
3            -999.000             -999.000             -999.000   
4            -999.000             -999.000             -999.000   

   PRI_jet_subleading_pt  PRI_jet_subleading_eta  PRI_jet_subleading_phi  \
0                 46.062                    1.24                  -2.475   
1               -999.000                 -999.00                -999.000   
2               -999.000                 -999.00                -999.000   
3               -999.000                 -999.00                -999.000   
4               -999.000                 -999.00                -999.000   

   PRI_jet_all_pt  
0         113.497  
1          46.226  
2          44.251  
3           0.000  
4           0.000  

[5 rows x 32 columns]

## Data preprocessing 

We notice that there are some rows that have -999 as value for some columns. This mean that the corresponding particle doesn't produce a result for the corresponding column. 
For example the collision of the particle 100002 doesn't provide DER_mass_MMC 

In [35]:
# we replace all -999
tX = np.where(tX == -999,np.nan,tX)
df.replace( -999,np.nan,inplace=True)
df.describe()

Id   DER_mass_MMC  DER_mass_transverse_met_lep  \
count  250000.000000  211886.000000                250000.000000   
mean   224999.500000     121.858528                    49.239819   
std     72168.927986      57.298157                    35.344886   
min    100000.000000       9.044000                     0.000000   
25%    162499.750000      91.885250                    19.241000   
50%    224999.500000     112.406000                    46.524000   
75%    287499.250000     135.482000                    73.598000   
max    349999.000000    1192.026000                   690.075000   

        DER_mass_vis       DER_pt_h  DER_deltaeta_jet_jet  DER_mass_jet_jet  \
count  250000.000000  250000.000000          72543.000000      72543.000000   
mean       81.181982      57.895962              2.403735        371.783360   
std        40.828691      63.655682              1.742226        397.699325   
min         6.329000       0.000000              0.000000         13.602000   
25%        59.388750      14.068750              0.882500        111.977000   
50%        73.752000      38.467500              2.107000        225.885000   
75%        92.259000      79.169000              3.690000        478.226000   
max      1349.351000    2834.999000              8.503000       4974.979000   

       DER_prodeta_jet_jet  DER_deltar_tau_lep     DER_pt_tot  ...  \
count         72543.000000       250000.000000  250000.000000  ...   
mean             -0.821688            2.373100      18.917332  ...   
std               3.584362            0.782911      22.273494  ...   
min             -18.066000            0.208000       0.000000  ...   
25%              -2.629000            1.810000       2.841000  ...   
50%              -0.244000            2.491500      12.315500  ...   
75%               0.958000            2.961000      27.591000  ...   
max              16.690000            5.684000    2834.999000  ...   

         PRI_met_phi  PRI_met_sumet    PRI_jet_num  PRI_jet_leading_pt  \
count  250000.000000  250000.000000  250000.000000       150087.000000   
mean       -0.010119     209.797178       0.979176           84.822105   
std         1.812223     126.499506       0.977426           60.662276   
min        -3.142000      13.678000       0.000000           30.000000   
25%        -1.575000     123.017500       0.000000           44.422500   
50%        -0.024000     179.739000       1.000000           65.561000   
75%         1.561000     263.379250       2.000000          103.342000   
max         3.142000    2003.976000       3.000000         1120.573000   

       PRI_jet_leading_eta  PRI_jet_leading_phi  PRI_jet_subleading_pt  \
count        150087.000000        150087.000000           72543.000000   
mean             -0.003275            -0.012393              57.679474   
std               1.784546             1.813385              31.985782   
min              -4.499000            -3.142000              30.000000   
25%              -1.342000            -1.584000              37.312000   
50%               0.000000            -0.033000              47.902000   
75%               1.336000             1.562000              66.637000   
max               4.499000             3.141000             721.456000   

       PRI_jet_subleading_eta  PRI_jet_subleading_phi  PRI_jet_all_pt  
count            72543.000000            72543.000000   250000.000000  
mean                -0.011845               -0.001582       73.064591  
std                  2.031743                1.816950       98.015662  
min                 -4.500000               -3.142000        0.000000  
25%                 -1.612000               -1.576500        0.000000  
50%                 -0.010000               -0.002000       40.512500  
75%                  1.589500                1.576000      109.933750  
max                  4.500000                3.142000     1633.433000  

[8 rows x 31 columns]

## We try with Stochastic gradient descent 

In [36]:
initial_w = np.zeros((tX.shape[1],1))
max_iters = 50
gamma = 0.6
weights , loss = least_squares_SGD(y,tX,initial_w,max_iters,gamma)

print(loss)

[autoreload of helpers failed: Traceback (most recent call last):
  File "C:\Users\jalel\Anaconda3\envs\ml\lib\site-packages\IPython\extensions\autoreload.py", line 245, in check
    superreload(m, reload, self.old_objects)
  File "C:\Users\jalel\Anaconda3\envs\ml\lib\site-packages\IPython\extensions\autoreload.py", line 434, in superreload
    module = reload(module)
  File "C:\Users\jalel\Anaconda3\envs\ml\lib\imp.py", line 314, in reload
    return importlib.reload(module)
  File "C:\Users\jalel\Anaconda3\envs\ml\lib\importlib\__init__.py", line 169, in reload
    _bootstrap._exec(spec, module)
  File "<frozen importlib._bootstrap>", line 630, in _exec
  File "<frozen importlib._bootstrap_external>", line 724, in exec_module
  File "<frozen importlib._bootstrap_external>", line 860, in get_code
  File "<frozen importlib._bootstrap_external>", line 791, in source_to_code
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "C:\Users\jalel\Desktop\9ray

MemoryError: 

## Visualization and exploring the dataset

In [13]:
df.replace(-999,0,inplace=True)
df.head()

Id Prediction  DER_mass_MMC  DER_mass_transverse_met_lep  DER_mass_vis  \
0  100000          s       138.470                       51.655        97.827   
1  100001          b       160.937                       68.768       103.235   
2  100002          b           NaN                      162.172       125.953   
3  100003          b       143.905                       81.417        80.943   
4  100004          b       175.864                       16.915       134.805   

   DER_pt_h  DER_deltaeta_jet_jet  DER_mass_jet_jet  DER_prodeta_jet_jet  \
0    27.980                  0.91           124.711                2.666   
1    48.146                   NaN               NaN                  NaN   
2    35.635                   NaN               NaN                  NaN   
3     0.414                   NaN               NaN                  NaN   
4    16.405                   NaN               NaN                  NaN   

   DER_deltar_tau_lep  ...  PRI_met_phi  PRI_met_sumet  PRI_jet_num  \
0               3.064  ...       -0.277        258.733            2   
1               3.473  ...       -1.916        164.546            1   
2               3.148  ...       -2.186        260.414            1   
3               3.310  ...        0.060         86.062            0   
4               3.891  ...       -0.871         53.131            0   

   PRI_jet_leading_pt  PRI_jet_leading_eta  PRI_jet_leading_phi  \
0              67.435                2.150                0.444   
1              46.226                0.725                1.158   
2              44.251                2.053               -2.028   
3                 NaN                  NaN                  NaN   
4                 NaN                  NaN                  NaN   

   PRI_jet_subleading_pt  PRI_jet_subleading_eta  PRI_jet_subleading_phi  \
0                 46.062                    1.24                  -2.475   
1                    NaN                     NaN                     NaN   
2                    NaN                     NaN                     NaN   
3                    NaN                     NaN                     NaN   
4                    NaN                     NaN                     NaN   

   PRI_jet_all_pt  
0         113.497  
1          46.226  
2          44.251  
3           0.000  
4           0.000  

[5 rows x 32 columns]

In [14]:
y

array([ 1., -1., -1., ...,  1., -1., -1.])

## Generate predictions and save ouput in csv format for submission:

In [6]:
DATA_TEST_PATH = '' # TODO: download train data and supply path here 
_, tX_test, ids_test = load_csv_data(DATA_TEST_PATH)

OSError:  not found.

In [ ]:
OUTPUT_PATH = '' # TODO: fill in desired name of output file for submission
y_pred = predict_labels(weights, tX_test)
create_csv_submission(ids_test, y_pred, OUTPUT_PATH)